### ISSUE TITLE
Segmentation fault in 3.22

### ISSUE THREAD

**Issue by Axelandre42** [libreswan/libreswan#152]
## The bug

I tried to install _libreswan_ from sources.
I compiled and installed using:
```
root@<host>:~# make programs
root@<host>:~# make install
```
Then I configured a new connection:
```
#
# /etc/ipsec.d/home-tunnel.conf
#

conn home-subnet
    also=home-tunnel
    leftsubnet=192.168.0.0/24
    auto=add

conn home-subnet6 # Unused at the moment
    also=home-tunnel
    connaddrfamily=ipv6
    leftsubnet=fda3:84eb:c4b7:aff6::/64
    auto=add

conn home-tunnel
    left=<ip>
    right=<ip>
    authby=secret
```
I wrote _pre-shared keys_ in the `/etc/ipsec.secrets` and I ran the software:
```
root@<host>:~# ipsec setup start
Redirecting to: systemctl start ipsec.service
Job for ipsec.service failed because a fatal signal was delivered to the control process.
See "systemctl status ipsec.service" and "journalctl -xe" for details.
```
So I checked the status of the service:
```
root@<host>:~# systemctl --full status ipsec.service
● ipsec.service - Internet Key Exchange (IKE) Protocol Daemon for IPsec
   Loaded: loaded (/lib/systemd/system/ipsec.service; enabled; vendor preset: enabled)
   Active: failed (Result: signal) since Wed 2018-01-17 14:59:28 UTC; 42s ago
     Docs: man:ipsec(8)
           man:pluto(8)
           man:ipsec.conf(5)
  Process: 3342 ExecStopPost=/usr/local/sbin/ipsec --stopnflog (code=exited, status=0/SUCCESS)
  Process: 3340 ExecStopPost=/sbin/ip xfrm state flush (code=exited, status=0/SUCCESS)
  Process: 3336 ExecStopPost=/sbin/ip xfrm policy flush (code=exited, status=0/SUCCESS)
  Process: 3333 ExecStop=/usr/local/libexec/ipsec/whack --shutdown (code=exited, status=1/FAILURE)
  Process: 3331 ExecStart=/usr/local/libexec/ipsec/pluto --leak-detective --config /etc/ipsec.conf --nofork (code=killed, signal=SEGV)
  Process: 3318 ExecStartPre=/usr/local/sbin/ipsec --checknflog (code=exited, status=0/SUCCESS)
  Process: 3314 ExecStartPre=/usr/local/sbin/ipsec --checknss (code=exited, status=0/SUCCESS)
  Process: 3137 ExecStartPre=/usr/local/libexec/ipsec/_stackmanager start (code=exited, status=0/SUCCESS)
  Process: 3134 ExecStartPre=/usr/local/libexec/ipsec/addconn --config /etc/ipsec.conf --checkconfig (code=exited, status=0/SUCCESS)
 Main PID: 3331 (code=killed, signal=SEGV)
    Tasks: 0 (limit: 4915)
   CGroup: /system.slice/ipsec.service

Jan 17 14:59:28 asterisk-host-fwaeles systemd[1]: Failed to start Internet Key Exchange (IKE) Protocol Daemon for IPsec.
Jan 17 14:59:28 asterisk-host-fwaeles systemd[1]: ipsec.service: Unit entered failed state.
Jan 17 14:59:28 asterisk-host-fwaeles systemd[1]: ipsec.service: Failed with result 'signal'.
Jan 17 14:59:28 asterisk-host-fwaeles systemd[1]: ipsec.service: Service hold-off time over, scheduling restart.
Jan 17 14:59:28 asterisk-host-fwaeles systemd[1]: Stopped Internet Key Exchange (IKE) Protocol Daemon for IPsec.
Jan 17 14:59:28 asterisk-host-fwaeles systemd[1]: ipsec.service: Start request repeated too quickly.
Jan 17 14:59:28 asterisk-host-fwaeles systemd[1]: Failed to start Internet Key Exchange (IKE) Protocol Daemon for IPsec.
Jan 17 14:59:28 asterisk-host-fwaeles systemd[1]: ipsec.service: Unit entered failed state.
Jan 17 14:59:28 asterisk-host-fwaeles systemd[1]: ipsec.service: Failed with result 'signal'.
```
It looked strange so I tried to run it using the command line:
```
user@host:~# ipsec pluto --leak-detective --config /etc/ipsec.conf --nofork
Segmentation fault
```
Then I used _GDB_ to check what was wrong:
```
root@<host>:~# gdb --args /usr/local/libexec/ipsec/pluto --leak-detective --config /etc/ipsec.conf --nofork
GNU gdb (Debian 7.12-6) 7.12.0.20161007-git
[...]
Type "apropos word" to search for commands related to "word"...
Reading symbols from /usr/local/libexec/ipsec/pluto...done.
(gdb) run
Starting program: /usr/local/libexec/ipsec/pluto --leak-detective --config /etc/ipsec.conf --nofork
[Thread debugging using libthread_db enabled]
Using host libthread_db library "/lib/aarch64-linux-gnu/libthread_db.so.1".

Program received signal SIGSEGV, Segmentation fault.
add_state_prefix (buf=0xffffff8000e0) at /usr/src/libreswan-3.22/programs/pluto/log.c:150
150     {
(gdb) list 150
145     {
146             lswlogf(buf, "%03d ", rc);
147     }
148
149     static void add_state_prefix(struct lswlog *buf)
150     {
151             if (!pthread_equal(pthread_self(), main_thread)) {
152                     return;
153             }
154
(gdb) quit
A debugging session is active.

        Inferior 1 [process 4810] will be killed.

Quit anyway? (y or n) y
```

That's the error...

## The config

Tested on both [Scaleway's BareMetal ARM servers](https://www.scaleway.com/) and [Rock64](https://www.pine64.org/?page_id=7147)

* System : __Debian (9) Stretch__
  * Kernel (Scaleway) : __Linux host 4.9.23-std-1 #1 SMP Mon Apr 24 13:18:14 UTC 2017 aarch64 GNU/Linux__
  * Kernel (Rock64) : __Linux host 4.4.77-rockchip-ayufan-118 #1 SMP Thu Sep 14 21:59:24 UTC 2017 aarch64 GNU/Linux__

**Comment by letoams** [comment]
On Wed, 17 Jan 2018, Alexandre Waeles wrote:

This is caused by a misconfiguration that tries to
create a conn where leftsubnet= and rightsubnet=
have a different address family (ipv4 vs ipv6)

We will look at better failing this problem :)

Paul

**Comment by letoams** [comment]
Fixed, will be in 3.27

**Comment by vitorhugop** [comment]
> On Wed, 17 Jan 2018, Alexandre Waeles wrote: This is caused by a misconfiguration that tries to create a conn where leftsubnet= and rightsubnet= have a different address family (ipv4 vs ipv6) We will look at better failing this problem :) Paul
> […](#)

Hello Paul!

This error can appear if a tunnel is started with leftsubnet configured and without rightsubnet?

Libreswan-3.23 and CentOS 7.

Thank you.

**Comment by letoams** [comment]
On Mon, 14 Jan 2019, vitorhugop wrote:

> This error can appear if a tunnel is started with leftsubnet configured and without rightsubnet?
> 
> Libreswan-3.23 and CentOS 7.

AFAIK, this was fixed in 3.25:

* pluto: Obsoleted connaddrfamily= (fixes 6in4 and 4in6) [Paul]
* pluto: New hostaddrfamily= and clientaddrfamily= (only needed w DNS) [Paul]
* pluto: Parser bug prevented 4in6 config [mhuntxu at github, Daniel M.  Weeks]

Could you re-test?

Paul
